portable/GCC/ARM_CM4F/port.c

```diff
@@ -607,26 +620,22 @@ void xPortSysTickHandler( void )
 			be, but using the tickless mode will inevitably result in some tiny
 			drift of the time maintained by the kernel with respect to calendar
 			time*/
-			portNVIC_SYSTICK_CTRL_REG = ( portNVIC_SYSTICK_CLK_BIT | portNVIC_SYSTICK_INT_BIT );
+			portNVIC_SYSTICK_CTRL_REG = ( portNVIC_SYSTICK_CLK_BIT_SETTING | portNVIC_SYSTICK_INT_BIT );

```

Comment:
Same here. When setting portNVIC_SYSTICK_CTRL_REG with this assignment, portNVIC_SYSTICK_COUNT_FLAG_BIT (bit-16) is cleared. Looks like the code always ends up in the else branch. 

```
portNVIC_SYSTICK_CTRL_REG &= ~portNVIC_SYSTICK_ENABLE_BIT;
```

Reply:
Note that portNVIC_SYSTICK_COUNT_FLAG_BIT is a read-only bit.  It is cleared by any *read* of the CTRL register and by any write to the current-value register.  The original code [here](https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/4b353bfd7aa009d553f31b4e47dfb49a9f3c75b1/portable/GCC/ARM_CM4F/port.c#L610) relies on this behavior to stop the systick (the second time) without modifying the count flag.  Also note perhaps more importantly that there is no change here in the compiled code from the original; the new symbol portNVIC_SYSTICK_CLK_BIT_SETTING has the same value as the old symbol portNVIC_SYSTICK_CLK_BIT.

Reply:
I understand your PR did not change the line you are referring to. But I doubt there was a bug at [the line](https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/4b353bfd7aa009d553f31b4e47dfb49a9f3c75b1/portable/GCC/ARM_CM4F/port.c#L610), so that the if path (line 619-640) is never taken. 

I'm referring to this Cortex M4 datasheet -- https://developer.arm.com/docs/dui0553/a/cortex-m4-peripherals/system-timer-systick
```SYST_CSR``` is RW, and ```COUNTFLAG``` is at bit 16. 

I setup a quick demo to prove this line  
```
portNVIC_SYSTICK_CTRL_REG = ( portNVIC_SYSTICK_CLK_BIT | portNVIC_SYSTICK_INT_BIT );
```
overwrites ```COUNTFLAG``` bit. And it seems to be the case. In below snapshot, I added line 603-608 and set breakpoint at 607. So if we ever trap at 607, we should also trap at line 630, if  ```SYST_CSR``` is read-only. Not quite the case. (I'm using Keil uVision v5.30. And a minor note is that the System Tick Timer window also in the snapshot is incorrectly reading ```COUNTFLAG``` bit. In this case we trapped at 607, but bit 16 is not set. I'm not sure if that contributed to "read-only" observation in your comment.. )


![Screen Shot 2020-06-01 at 11 39 00 PM](https://user-images.githubusercontent.com/10982575/83488091-176ca300-a461-11ea-9076-a01b859fa86f.png)

Reply:
No, there's no bug here in the original code or in the proposed code.  And the "minor note" you mentioned about your test code incorrectly reading `COUNTFLAG` is not actually happening either.

Here is the explanation:  The `COUNTFLAG` is cleared by any **read** of the `SYS_CSR`.  So as when your new "if" statement executes, regardless of the outcome of that statement, it clears the COUNTFLAG.

See here:
![Capture](https://user-images.githubusercontent.com/61161750/83521904-beb6fd80-a494-11ea-832b-68a876e47327.PNG)

Reply:
I'm good now, apologies for the detour. Also break at if statement instead of inside the if, and occasionally I do see the bit got set and cleared by read. Going to resolve the this conversation and ```COUNTFLAG``` related.

Reply:
No worries.  It's a strange bit.  I appreciate the thorough review.